tls/client-tls-pkcallback.c

```diff
@@ -0,0 +1,367 @@
+static int load_file(const char* fname, byte** buf, size_t* bufLen)
+{
+    int ret;
+    long int fileSz;
+    XFILE lFile;
+
+    if (fname == NULL || buf == NULL || bufLen == NULL)
+        return BAD_FUNC_ARG;
+
+    /* set defaults */
+    *buf = NULL;
+    *bufLen = 0;
+
+    /* open file (read-only binary) */
+    lFile = XFOPEN(fname, "rb");
+    if (!lFile) {
+        printf("Error loading %s\n", fname);
+        return BAD_PATH_ERROR;
+    }
+
+    fseek(lFile, 0, SEEK_END);
+    fileSz = (int)ftell(lFile);
+    rewind(lFile);
+    if (fileSz  > 0) {
+        *bufLen = (size_t)fileSz;
+        *buf = (byte*)malloc(*bufLen);
+        if (*buf == NULL) {
+            ret = MEMORY_E;
+            printf("Error allocating %lu bytes\n", (unsigned long)*bufLen);
+        }
+        else {
+            size_t readLen = fread(*buf, *bufLen, 1, lFile);
+
+            /* check response code */
+            ret = (readLen > 0) ? 0 : -1;
+        }
+    }
+    else {
+        ret = BUFFER_E;
+    }
+    fclose(lFile);
+
+    return ret;
+}
+
+static int load_key_file(const char* fname, byte** derBuf, word32* derLen)
+{
+    int ret;
+    byte* buf = NULL;
+    size_t bufLen;
+
+    ret = load_file(fname, &buf, &bufLen);
+    if (ret != 0)
+        return ret;
+
+    *derBuf = (byte*)malloc(bufLen);
+    if (*derBuf == NULL) {
+        free(buf);
+        return MEMORY_E;
+    }
+
+    ret = wc_KeyPemToDer(buf, (word32)bufLen, *derBuf, (word32)bufLen, NULL);
+    if (ret < 0) {
+        free(buf);
+        free(*derBuf);
+        return ret;
+    }
+    *derLen = ret;
+    free(buf);
+
+    return 0;
+}
+
+static int myEccSign(WOLFSSL* ssl, const byte* in, word32 inSz,
+        byte* out, word32* outSz, const byte* key, word32 keySz, void* ctx)
+{
+    int ret;
+    byte* keyBuf = (byte*)key;
+    PkCbInfo* cbInfo = (PkCbInfo*)ctx;
+
+    printf("PK ECC Sign: inSz %u, keySz %u\n", inSz, keySz);
+
+#ifdef WOLFSSL_ASYNC_CRYPT
+    if (cbInfo->state == 0) {
+        cbInfo->state++;
+        printf("PK ECC Sign: Async Simulate\n");
+        return WC_PENDING_E;
+    }
+#endif
+
+    ret = load_key_file(cbInfo->keyFile, &keyBuf, &keySz);
+    if (ret == 0) {
+        ret = wc_ecc_init(&cbInfo->key);
+        if (ret == 0) {
+            word32 idx = 0;
+            ret = wc_EccPrivateKeyDecode(keyBuf, &idx, &cbInfo->key, keySz);
+            if (ret == 0) {
+                WC_RNG *rng = wolfSSL_GetRNG(ssl);
+
+                printf("PK ECC Sign: Curve ID %d\n", cbInfo->key.dp->id);
+                ret = wc_ecc_sign_hash(in, inSz, out, outSz, rng, &cbInfo->key);
+            }
+            wc_ecc_free(&cbInfo->key);
+        }
+    }
+    free(keyBuf);
+
+#ifdef WOLFSSL_ASYNC_CRYPT
+    cbInfo->state = 0;
+#endif
+
+    printf("PK ECC Sign: ret %d outSz %u\n", ret, *outSz);
+
+    return ret;
+}
+#endif /* HAVE_PK_CALLBACKS */
+
+int main(int argc, char** argv)
+{
+    int                ret = 0, err;
```

Comment:
I think ret should be initialized to -1

Reply:
Removed, since it is always set below.

Reply:
No...it isn't.  See lines 236 and 249. It's easy to miss.  Therefore, I'd suggest initializing to -1 and then removing all the places where it gets set to -1 later.